fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/utils/FedoraTypesUtils.java

```diff
@@ -329,22 +337,67 @@ public static Node getJcrNode(final FedoraResource resource) {
     }
 
     /**
-     * Update the fedora:lastModified date of the ldp:membershipResource if this node is a direct
-     * or indirect container.
+     * Given a JCR Node, fetch the parent's ldp:insertedContentRelation value, if
+     * one exists.
+     *
+     * @param node the JCR Node
+     * @return the ldp:insertedContentRelation Resource, if one exists.
+     */
+    public static Optional<Resource> ldpInsertedContentProperty(final Node node) {
+        try {
+            if (node.getDepth() > 0) {
+                final Node parent = node.getParent();
+
+                if (parent.hasProperty(LDP_MEMBER_RESOURCE) && parent.isNodeType(LDP_INDIRECT_CONTAINER) &&
+                        parent.hasProperty(LDP_INSERTED_CONTENT_RELATION)) {
+                    return Optional.of(createResource(parent.getProperty(LDP_INSERTED_CONTENT_RELATION).getString()));
+                }
+            }
+        } catch (final RepositoryException ex) {
+            throw new RepositoryRuntimeException(ex);
+        }
+        return empty();
+    }
+
+    public static Function<Session, Function<Resource, Optional<String>>> resourceToProperty = session -> resource -> {
```

Comment:
This might be clearer as a method:
```
public static Function<Resource, Optional<String>> resourceToProperty(Session session) {
```
1. The only time you use it that I see, you explicitly call `apply`. That's the very semantic of a plain method, which is more concise.
2. The `Session` is actually part of the state of the `Resource -> Optional<String>` function.